### DISCUSSION TITLE
Replace redirect after a fetcher submit

### DISCUSSION THREAD

**Discussion by mattoni** [remix-run/remix#2083] [Q&A] [answered]
I've got a modal that appears when a certain search parameter is populated. The modal has a form that submits some data to an action api route via a fetcher, and that route returns a redirect. Is it possible to make it so that the back button doesn't return the user to the modal in this scenario? Adding a replace prop onto the fetcher.Form didn't seem to work.

Thanks!

**Comment by silvenon** [comment]
Fetcher doesn't navigate, so I'm a little confused by your question. It's expected that replace wouldn't work, because there was no navigation to begin with. Although I haven't tried myself what happens when you submit via fetcher to an action that returns a redirect.

Would submitting via `useSubmit` make sense in your situation? In that case you could pass `replace: true`.

**Reply by mattoni** [reply]
Fetcher does navigate if the request returns a redirect, so I think there should be a way to handle that/have it respect the place on the form.

**Comment by ryanflorence** [comment, marked as answer]
Any reason you're using a fetcher instead of `<Form>` or `submit()`?

**Reply by mattoni** [reply]
Three reasons - I need this modal to be shareable via URL, it needs to appear over top of several different pages, and it needs to reload the loaders on submit from a higher up parent in the route.

The navigation is desired on a successful submit to close out the modal (remove the ?search param). If I made this modal a dedicated route, I'd need to repeat that all over the code as a nested route for any page it needs to appear 'on top of'. Search parameters made the most sense here. 

Without the redirect to close the modal, the root loader never reloads which is problematic as this modal is a form that does mutations to some data higher up.

The 'useSubmit' may work, but then it seems like I'm breaking out of the 'remix way' when everything except the replace functionality works well with a fetcher + redirect.

**Reply by mattoni** [reply]
Though, I may be over-thinking this. using a regular <Form> may work in this context. Let me give it a shot.

**Reply by ryanflorence** [reply]
Redirect from the route you're posting to and you don't have to repeat any code. Every `<Form replace action="/modal/submit/endpoint" />` will redirect to the page it came from, either from the `request.headers.get("Referer")` or put it explicitly in the form:

```jsx
let location = useLocation();

<Form replace action="/modal/action">
  <input type="hidden" name="redirectTo" value={location.pathname} />
// ...
```

**Reply by ryanflorence** [reply]
that said, I'm honestly not sure what `<fetcher.Form replace>` even does with the history stack right now.

**Reply by mattoni** [reply]
Awesome, I think I was thinking about it too much like the "usual way" with react. Using a regular form achieves the goal here. Thanks so much!

**Reply by mattoni** [reply]
One followup though, if the form errors out, I get an error `Route XXX has no component! Please go add a 'default' export in the route module file. If you were trying to navigate or submit to a resource route, use '<a>' instead of '<Link>' or '<Form reloadDocument>'.`

Not sure what to do in this situation? The route specified is the one my form posts to, which doesn't have a component attached to it. It's returning a 422 status with a json body.

**Reply by mattoni** [reply]
I guess this is what made me lean toward the fetcher - I need to return errors from the post without redirecting....but I want to redirect on success. The fetcher accomplishes this whereas regular forms still redirect to the original route even if the response isn't 2xx. 

Is this expected behavior?

**Reply by silvenon** [reply]
This is exactly why I reached for `useFetcher`.

`<Form action="/api">` always navigates to `/api`, so I think the reason for the limitation you describe is because otherwise the workflows with and without JavaScript would diverge, i.e. it wouldn't work without JavaScript. However, I'm getting the error you mentioned even regardless of what the `action` in the `/api` returns, but that's consistent with the same reason — without JavaScript you would lose the data returned by the action.

`<fetcher.Form replace>` would be weird because fetcher itself doesn't navigate, only with redirect. Instead of redirect I recommend using `useNavigate` + `replace: true`, considering that you're in client-side land anyway.
